go/database/mpt/verification_test.go

```diff
@@ -412,6 +414,154 @@ func TestVerification_ValueNodeHashModificationIsDetected(t *testing.T) {
+	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+		ctrl := gomock.NewController(t)
+		observer := NewMockVerificationObserver(ctrl)
+
+		testHash := common.Keccak256([]byte{1})
+		codes := map[common.Hash][]byte{
+			testHash: {1},
+		}
+
+		gomock.InOrder(
+			observer.EXPECT().StartVerification(),
+			observer.EXPECT().Progress("Obtaining read access to files ..."),
+			observer.EXPECT().Progress(fmt.Sprintf("Checking contract codes ...")),
+			observer.EXPECT().Progress(fmt.Sprintf("There are %d contracts not referenced by any accounts:", len(codes))),
+			observer.EXPECT().Progress(fmt.Sprintf("%x\n", testHash)),
+			observer.EXPECT().Progress(gomock.Any()).MinTimes(1),
+			observer.EXPECT().EndVerification(nil),
+		)
+
+		if err := writeCodes(codes, dir+"/codes.dat"); err != nil {
+			t.Fatalf("failed to write code file")
+		}
+
+		if err := VerifyMptState(dir, config, roots, observer); err != nil {
+			t.Errorf("found unexpected error in fresh forest: %v", err)
+		}
+	})
+}
+
+func TestVerification_UnreadableCodesReturnError(t *testing.T) {
+	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+		f, err := os.OpenFile(filepath.Join(dir, "codes.dat"), os.O_APPEND|os.O_WRONLY, os.ModeAppend)
+		if err != nil {
+			t.Fatalf("failed to open codes file: %v", err)
+		}
+
+		_, err = f.Write([]byte{1})
+		if err != nil {
+			t.Fatalf("failed to open write to codes file: %v", err)
+		}
+
+		if err = f.Close(); err != nil {
+			t.Fatalf("failed to close codes file: %v", err)
+		}
+
+		if err = VerifyMptState(dir, config, roots, NilVerificationObserver{}); err == nil {
+			t.Errorf("different hash in code file should have been detected")
+		}
+	})
+}
+
+func TestVerification_PassingNilAsObserverDoesNotPanic(t *testing.T) {
+	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+		if err := VerifyMptState(dir, config, roots, nil); err != nil {
+			t.Errorf("found unexpected error in verification: %v", err)
+		}
+	})
+}
+
+func TestVerifyFileForest_PassingNilAsObserverDoesNotPanic(t *testing.T) {
+	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+		if err := VerifyFileForest(dir, config, roots, nil); err != nil {
+			t.Errorf("found unexpected error in verification: %v", err)
+		}
+	})
+}
+func TestVerification_CodesAreNotCheckedTwice(t *testing.T) {
+	runVerificationTest(t, func(t *testing.T, dir string, config MptConfig, roots []Root) {
+		testHash := common.Keccak256([]byte{1})
+		codes := map[common.Hash][]byte{
+			testHash: {1},
+		}
+		if err := writeCodes(codes, filepath.Join(dir, "codes.dat")); err != nil {
+			t.Fatalf("failed to write code file")
+		}
+
+		encoder, _, _, _ := getEncoder(config)
+
+		// modify two different nodes to append same hash to two accounts
+		modifyFirstNode(t, filepath.Join(dir, "accounts"), encoder, func(node *AccountNode) {
```

Comment:
I have always created tests this way. Imo it removes redundancy and it also applies a clear thought how the data is incorrect right in the test.
If you still feel I should create a new data set, even tho its only for one test I will do it.

Reply:
I agree with what you are saying, but I do not think it is what is actually happening in the tests.  It is alright to have a correct data and then corrupt them on purpose to mimic an erroneous state. 

But here it is:  all test accounts have empty code-hash. It is not because it represents an error-free state, it is just a coincidence that these code-hashes were not used in tests before. 

you now used these tests and inserted code-hashes in the tests. This I think is wrong, the test data should be created ready to use. I think that the hashes should be created right where the data are created. I guess you do not need to create a separate test data, just insert the code hashes where the test data are created. You can have a list of test byte-code sequences and their hashes initialised in advance. 

finally you create codes.dat, where you insert data that do not match accounts. This is where you do on purpose manipulation with data to create an erroneous state - and this is a correct way. 

if you do it this way, you can revert all the newly introduced methods to modify first/last node, etc. 

btw: it is happening I think 3-times, not only in this test.

Reply:
I see, you dont like appending `emptyHash` which actually skips the validation and then corrupting the data. Make sense. Thanks

Reply:
Please have a look at this commit: https://github.com/Fantom-foundation/Carmen/pull/881/commits/33c354abe903fa6afa3efe5525a0ad85e3400d51 

I hope it explains better what I meant. 

It breaks one test, but this test was somehow unclear, because I do not see how does it test that a code is not checked twice. Can you please have a look and fix?

Reply:
Thanks for the help. I was implementing something very similar.